### PR TITLE
Compatibility with Android SDK < 24

### PR DESCRIPTION
**Pull request by masc3d**
resolving #774

### CODE REVIEW

**Comment by velo** [comment]
Sounds fair, but CI is not happy about it

**Comment by masc3d** [comment]
@velo @kdavisk6 as suspected, here comes the rabbit hole (eg `String.join`)

**Comment by masc3d** [comment]
I would repackage predicates, `String.join` and any classes required for compat in the future (and there will be more) from openjdk into `feign.util`. is this ok with you?

**Comment by masc3d** [comment]
can't do this without relaxing license header restrictions though

**Comment by masc3d** [comment]
I also see `feign-10` makes extended use of `java.beans` and `hamcrest` all of which are not compatible with android. opinions on resolutions?

**Comment by masc3d** [comment]
the way I see it we also need safeguards (eg. lint) to prevent potential android breakage on every minor change.

**Comment by masc3d** [comment]
> I agree with @velo. Those implementations are quite simple, so we can add our own.

not applicable. this is a rabbit hole and more and more compat code will be added if android should be properly supported. repackaging is probably the only reasonable option.

**Comment by velo** [comment]
What do you mean by `repackaging`?

**Comment by masc3d** [comment]
> What do you mean by `repackaging`?

(very) selectively import classes from `streamsupport` and possibly `openjdk` (eg. `StringJoiner` including static utility methods) into a dedicated package `feign.java8` which has relaxed license header rules.

**Comment by velo** [comment]
Ok, `streamsupport` is acceptable (personal opinion here) now, `apache-commons` no way jose

**Comment by masc3d** [comment]
> `apache-commons` no way jose
not required when repackaging eg. `StringJoiner` from openjdk

File: core/src/main/java/feign/java8/util/StringJoiner.java [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+ * This code is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version 2 along with this work;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA or visit www.oracle.com
```
Comment (velo) [inline core/src/main/java/feign/java8/util/StringJoiner.java]: @kdavisk6 thoughts?

Comment (kdavisk6) [reply core/src/main/java/feign/java8/util/StringJoiner.java]: I don't think this is appropriate.  I would prefer that we implement our own version of the `join` function over copying code like this.

Regarding the problem at hand, I'm still not comfortable with creating our own "polyfills" when the JDK supports these natively.

Comment (masc3d) [reply core/src/main/java/feign/java8/util/StringJoiner.java]: >  I would prefer that we implement our own version of the join
@kdavisk6 if it was just about `join` I would agree

but in the end any solution will be kind of awkward. I already detected more incompatibilities.
maybe we should conclude that `feign-10` can only be compatible with SDK >= 24.

**Comment by kdavisk6** [comment]
Based on our conversations, I'm going to close this PR in favor of documenting that, if you need to work on Android versions less than 24, that you use the Feign 9.x releases.

**Pull request closed without merging**